aws-sdk-resources/spec/services/s3/bucket/delete_spec.rb

```diff
@@ -34,13 +34,119 @@ module S3
                 { key: 'key', version_id: 'vid-5' },
               ])
             }
-          )
+          }
+        end
 
-          expect(client).to receive(:delete_bucket).with(bucket: 'bucket-name')
+        context 'the bucket deletes without failure' do
+          it 'deletes object versions from the bucket and then deletes the bucket' do
+            client
+              .stub_responses(
+                :list_object_versions,
+                list_object_versions_request
+              )
 
-          bucket.delete!
+            expect(client)
+              .to receive(:delete_objects)
+              .with(delete_bucket_request)
+
+            expect(client).to receive(:delete_bucket).with(bucket: 'bucket-name')
+
+            bucket.delete!
+          end
         end
 
+        context 'the bucket raises a Errors::BucketNotEmpty exception' do
+          let(:bucket_not_empty_exception) do
+            Errors::BucketNotEmpty.new nil, ''
+          end
+
+          it 'retries' do
+            client
+              .stub_responses(
+                :list_object_versions,
+                list_object_versions_request
+              )
+
+            expect(client)
+              .to receive(:delete_objects)
+              .with(delete_bucket_request)
+              .exactly(3).times
+
+            expect(client).to receive(:delete_bucket)
+              .with(bucket: 'bucket-name')
+              .and_raise(bucket_not_empty_exception).ordered
+
+            expect(Kernel).to receive(:sleep).with(1.3).ordered
+
+            expect(client).to receive(:delete_bucket)
+              .with(bucket: 'bucket-name')
+              .and_raise(bucket_not_empty_exception).ordered
+
+            expect(Kernel).to receive(:sleep).with(1.6900000000000002).ordered
```

Comment:
Add a comment around this to explain the magic number, at least. That way, it's a less confusing failure if the defaults are changed later. Or, move out to variables that are grouped together.